Issue#68 translate tutorials (#120)
reactjs/pt-BR.reactjs.org: content/tutorial/tutorial.md

![React Devtools](../images/tutorial/tictac-numbers.png)

**[Veja o código até este momento](https://codepen.io/gaearon/pen/aWWQOG?editors=0010)**

Parabéns! Você acabou de passar um "prop" de um componente pai Board para um componente filho Square. Passar props é a forma como os dados fluem em aplicações React, de pais para filhos.

> ### Fazendo um componente interativo {#making-an-interactive-componente}

Reviewer: ```suggestion
### Fazendo um componente interativo {#making-an-interactive-component}
```

Os Id's jamais devem ser alterados 

cc: @raulfdm

Author: Nice catch. Eu fiz um global search e fui substituindo... acho que passou sem querer.

3. Clique em "Change View" e escolha "Debug mode".
4. Na nova aba que se abre, o devtools deve agora ter uma aba React.

## Completando o jogo {#completing-the-game}

Agora nós temos os blocos para o nosso jogo da velha. Para completar o jogo, nós precisamos preencher os "X"s e os "O"s no tabuleiro e alguma maneira de determinar o vencedor.

### Movendo o state para cima {#lifting-state-up}

Atualmente, cada component Quandrado (Square) mantem o estado do jogo. Para checar o vencedor, nós vamos manter o valor de cada um dos 9 quadrados em uma posição.

Podemos pensar que o Tabuleiro (Board) poderia apenas perguntar para cada Quadrado pelo seu estado. Apesar desse modelo ser possível no React, nós o desencorajamos, pois, o código se torna muito difícil de ser compreendido, suscetível à bugs e difícil de refatorar. Ao invés disso, a melhor opção é guardar o estado do jogo no component pai (Tabuleiro). Este, irá dizer para cada Quadrado o que deverá ser exibido via prop, [assim como fizemos quando passamos o número de cada Quadrado](#passing-data-through-props).

> **Para coletar dados de múltiplos filhos (children), ou para fazer dois filhos se comunicarem entre si, você precisa declarar um estado compartilhado em seu component pai. O componente pai pode passar o estado de volta para os filhos através do uso de propriedades (props); isso mantém o componentes filhos em sincronia com os seus irmãos e também com o pai.**

Reviewer 2: Aqui foi utilizado "component" ao invés de componente, que acredito ser o padrão

> Criar estado em um component Pai é bem comum quando componentes React são refatorados -- Vamos aproveitar essa oportunidade para tentar o conceito na prática. Vamos adicionar um construtor no Tabuleiro e definir que seu estado inicial irá ter um array com 9 poisções preenchidas por nulo (null). Esses 9 nulls corresponderá aos 9 quadrados: 

Reviewer 2: Uso de "component"
Typo: poisções
E também:
> Esses 9 nulls corresponderá 

Nao sei se aqui seria melhor usar _nulls_ ou _nulos_ mas de qualquer forma a conjugação de _corresponderá_ esta inadequada

Reviewer 2: Não encontrei o padrão sobre o _null_ no glossário... thoughts?

Author: Então, sobre `null` eu sempre fico na dúvida. Querendo ou não é uma reserved word em JS. Na dúvida eu deixei ai, mas precisaria da opinião de mais alguém pra saber se é ok.

@halian-vilela

Reviewer 3: @raulfdm @Luanf 

Acho que do jeito que fizeram tá bom, dependendo de como aparece no texto por ser importante botar um "(nulo)" para dar mais contexto. Mas no geral, por ser um termo corriqueiro e uma palavra reservada, está OK manter em inglês mesmo.

  'O', null, null,
]
```
O método `renderSquare` do Tabuleiro atualmente está definido como:

```javascript
  renderSquare(i) {
    return <Square value={i} />;
  }
```

No começo, nós [passamos o seu valor (value) como prop](#passing-data-through-props) para o Tabuleiro mostrar números de 0 a 8 em cada Quadrado. Em outro passo anterior, nós trocamos os números com pela letra "X" [determinado no próprio estado do Quadrado](#making-an-interactive-component). Isso porque atualmente o Quadrado ignora o valor (`value`) recebido do Tabuleiro.

> Iremos agora utilizar o mesmo mecanismo de propriedades novamente. Vamos modificar o Tabuleiro para instruir cada Quadrado invidiualmente qual é o valor correto (`'X'`, `'O'` ou `null`). Nós já temos definidos o array de `quadrados` no construtor do Tabuleiro e iremos modificar o método `renderSquare` para definir o valor a partir do estado:

Reviewer 2: Acredito que 
> Iremos agora utilizar o mesmo mecanismo de propriedades novamente

Ficaria mais claro como
> Iremos agora utilizar _novamente_ o mesmo mecanismo de propriedades

  }
```

>Nota
>
>Nós quebrados o retorno do elemento em várias linhas para melhorar a legibilidade e adicionamos parentesis para que o Javascript não insira ponto e virgula após o `return` e quebre o código

Agora nós iremos passar duas props do Tabuleiro para o Quadrado: `value` e `onClick`. A propriedade `onClick` é uma função que será chamada quando o Quadrado for clicado. Nós manteremos as seguintes mudanças no component Quadrado:

* Substituir `this.state.value` por `this.props.value` no método `render`;
* Substituir `this.setState()` por `this.props.onClick()` no método `render`;

> * Deletar o `constructor` do Quadrado por não manteremos mais o state do jogo nele;

Reviewer 2: A frase 
> ... do Quadrado por não _manteremos_ mais o state do jogo nele

Esta equivocada; uma sugestão seria:

> do Quadrado, já que não manteremos mais o state do jogo nele;

Ou talvez apenas trocar o "por" por "pois"...

> Após essas mudanças, o component Quadrado se parecerá com isto:

Reviewer 2: Uso de "component"

}
```

**[Veja o código até este momento](https://codepen.io/gaearon/pen/ybbQJX?editors=0010)**

> Após essas mudanças, seremos capazes novamente de clicar nos Squares para preenche-los. Entretanto, agora o state é guardado no component Board ao invés de em cada Square. Quando o state do Board for alterado, os componentes Square serão re-renderizados automaticamente. Manter o state de todos os quadrados no component Board nos permitirá determinar o vencedor no futuro.

Reviewer 2: Uso de "component"

> Como o component Square não mantém mais state, os componentes Square receberão os valores do Board e o informarão quando forem clicados. Em "termos React", os Squares são agora **componentes controlados** (**controlled components**). O Board terá controle total sobre eles.

Reviewer 2: Uso de "component"

> Note como na função `handleClick`, nós chamados `.slick()` para criar uma cópia do array de quadrados para o modificar ao invés de faze-lo no array existente. Explicaremos o motivo quando criarmos uma copia do do array de `quadrados` na próxima sessão.

Reviewer 2: A função `.slick()` possui um nome bem legal mas ela ainda não existe em Javascript! :grin:
Brincadeiras a parte houve um typo na `.slice()`.

Alem disso algo nessa frase me parece confuso
>...para criar uma cópia do array de quadrados para o modificar ao invés de faze-lo no array existente.

Mas não tenho certeza do que. Fica a sugestão de talvez tentar outra construção. 
Mas de qualquer forma acredito que _faze-lo_ possui acento: _fazê-lo_

E por fim _copia_ deve ter um acento nessa seguinte frase, e "do" foi escrito duas vezes:
> ... quando criarmos uma copia do do array de ...

}
```

Agora, temos que mudar `this.props` para `props` nas duas ocorrências.

**[Veja o código até este momento](https://codepen.io/gaearon/pen/QvvJOv?editors=0010)**

>Nota
>

> >Quando modificamos o Square para ser um componente de função, nós também mudamos o `onClick={() => this.props.onClick()}` para uma versão mais curta `onClick={props.onClick}` (Note a falta de parentesis de ambos os lados). Em uma classe, nós usamos arrow function para acessar corretamente o valor do `this`, mas em um componente de função nós não precisamos nos preocupar com o `this`.

Reviewer 2: Acentuação:
```suggestion
>Quando modificamos o Square para ser um componente de função, nós também mudamos o `onClick={() => this.props.onClick()}` para uma versão mais curta `onClick={props.onClick}` (Note a falta de parêntesis de ambos os lados). Em uma classe, nós usamos arrow function para acessar corretamente o valor do `this`, mas em um componente de função nós não precisamos nos preocupar com o `this`.
```

  }
```

> Cada vez que o jogar faz uma jogada, `xIsNext` (um booleano) vai ser trocado para determinar qual jogador será o próximo e o state do jogo será salvo. Nós vamos atualizar a função `handleClick` do component Board para trocar o valor de `xIsNext`:

Reviewer 2: ```suggestion
Cada vez que o jogador fizer uma jogada, `xIsNext` (um booleano) vai ser trocado para determinar qual jogador será o próximo e o state do jogo será salvo. Nós vamos atualizar a função `handleClick` do component Board para trocar o valor de `xIsNext`:
```

  }
```

> Com essa alteração, "X"s e "O"s podem revezar. Vamos também alterar o texto "status" no render do Board para que seja exibido qual é de quem é a vez de jogar:

Reviewer 2: Pequena confusão:
```suggestion
Com essa alteração, "X"s e "O"s podem revezar. Vamos também alterar o texto "status" no render do Board para que seja exibido de quem é a vez de jogar:
```

}
```

**[Veja o código até este momento](https://codepen.io/gaearon/pen/KmmrBy?editors=0010)**

### Declarando o Vencedor {#declaring-a-winner}

> Agora que nós sabemos quem é o próximo a jogar, também mostraremos quando o jogo foi vencido e que não há mais turnos. Nós podemos determinar o vencedor adicionar uma função auxiliadora no fim do arquivo:

Reviewer 2: ```suggestion
Agora que nós sabemos quem é o próximo a jogar, também mostraremos quando o jogo foi vencido e que não há mais turnos. Nós podemos determinar o vencedor adicionando uma função auxiliadora no fim do arquivo:
```

  }
```

**[Veja o código até este momento](https://codepen.io/gaearon/pen/LyyXgK?editors=0010)**

Parabéns! Agora você tém um jogo da velha funcionando. E você também acabou de aprender o básico de React. *Você é* provavelmente o verdadeiro vencedor! 

## Adicionando Viagem no Tempo {#adding-time-travel}

Como exercício final, vamos tornar possível "voltar no tempo" para a jogada anterior no jogo.

### Armazenando o Histórico das jogadas {#storing-a-history-of-moves}

Se você mutou o array de `quadrados`, implementar viagem no tempo vai ser bem difícil.

> Entretanto, nós usamos `slice()` para criar uma nova cópia do array de `quadrados` após cada jogada e o [tratamos como imutável](#why-immutability-is-important). Isso vai nos possibilidade para armazenar cada versão anterior do array de `quadrados` e navegar através dos turnos que já aconteceram.

Reviewer 2: ```suggestion
Entretanto, nós usamos `slice()` para criar uma nova cópia do array de `quadrados` após cada jogada e o [tratamos como imutável](#why-immutability-is-important). Isso vai nos possibilitar armazenar cada versão anterior do array de `quadrados` e navegar através dos turnos que já aconteceram.
```

> Vamos armazenar cada array passado em um novo array chamado `history`. O array `history` representa todo os estados do tabuleiro, da primeira jogada até a última e algo como:

Reviewer 2: ```suggestion
Vamos armazenar cada array passado em um novo array chamado `history`. O array `history` representa todo os estados do tabuleiro, da primeira jogada até a última e é algo como:
```

]
```

Agora precisamos decidir a qual componente o state `history` pertencerá.

### Criando o estado, de novo {#lifting-state-up-again}

Nós vamos querer que o componente mais alto do jogo (Game) exiba a lista das jogadas passadas. Ele precisará de acesso ao `history` para fazer isso, então nós vamos criar o state `history` nele.

> Colocar `history` no componente Game nos permite remover o state dos `squares` para o seu component filho, Board.

Reviewer 2: Uso de "component"
```suggestion
Colocar `history` no componente Game nos permite remover o state dos `squares` para o seu componente filho, Board.
```

}
```

> Em seguida, vamos fazer o componente Board rebecer as propriedades `squares` e `onClick` do componente Game. Já de agora temos uma única funcão handler no Board para vários Squares, vamos precisar passar a localização de cada Square para o handler `onClick` para indicar qual quadrado foi clicado. A seguir os passos necessários para alterar o componente Board:

Reviewer 2: ```suggestion
Em seguida, vamos fazer o componente Board receber as propriedades `squares` e `onClick` do componente Game. Já que agora temos uma única funcão handler no Board para vários Squares, vamos precisar passar a localização de cada Square para o handler `onClick` para indicar qual quadrado foi clicado. A seguir os passos necessários para alterar o componente Board:
```

  }
```

>Nota
>
>Ao contrário do método `push()` em array que você já deve conhecer, o método `concat()` não muta o array origina, assim, preferimos ele.

> Neste ponto, o componente Board precisa somete dos métodos `renderSquare` e `render`. O estado do jogo e o método `handleClick` devem ficar no componente Game.

Reviewer 2: ```suggestion
Neste ponto, o componente Board precisa somente dos métodos `renderSquare` e `render`. O estado do jogo e o método `handleClick` devem ficar no componente Game.
```

**[Veja o código até este momento](https://codepen.io/gaearon/pen/EmmOqJ?editors=0010)**

### Mostrando os movimentos passados {#showing-the-past-moves}

> Já nós gravamos o histórico do jogo da velha, nós podemos mostrar ao jogador a lista das jogadas anteriores.

Reviewer 2: ```suggestion
Já que nós gravamos o histórico do jogo da velha, nós podemos mostrar ao jogador a lista das jogadas anteriores.
```

  }
```

**[Veja o código até este momento](https://codepen.io/gaearon/pen/EmmGEa?editors=0010)**

> Para cada jogada no histórico do nosso jogo da velha, nós criamos uma lista de items `<li>` que contém um botão `<button>`. O botão possui o handler `onClick` que chama o método `this.jumpTo()`. Nós não vamos implementar o método `jumpTo()` ainda. Por hora, vamos exibir a lista de jogadas que ocorreram no jogo e imprimir no console da ferramenta de desenvolvedor a mensagem: 

Reviewer 2: ```suggestion
Para cada jogada no histórico do nosso jogo da velha, nós criamos uma lista de itens `<li>` que contém um botão `<button>`. O botão possui o handler `onClick` que chama o método `this.jumpTo()`. Nós não vamos implementar o método `jumpTo()` ainda. Por hora, vamos exibir a lista de jogadas que ocorreram no jogo e imprimir no console da ferramenta de desenvolvedor a mensagem: 
```

>  Atenção:
>  Cada filho no array ou iterator deve ter uma propriedade "key" única. Verifique no método render do "Game"

Vamos discutir o que o alerta acima significa.

### Picking a Key {#picking-a-key}

> Quando renderizamos uma lista, o React guarda algumas informações sobre cada item renderizado da lista. Já quando atualizamos uma lista, ele precisa determinar o que mudou. Nós podemos ter adicionado, removido, reorganizado ou atualizado a lista de items.

Reviewer 2: ```suggestion
Quando renderizamos uma lista, o React guarda algumas informações sobre cada item renderizado da lista. Já quando atualizamos uma lista, ele precisa determinar o que mudou. Nós podemos ter adicionado, removido, reorganizado ou atualizado a lista de itens.
```

Imagine a transição de

```html
<li>Alexa: 7 tasks left</li>
<li>Ben: 5 tasks left</li>
```

Para

```html
<li>Ben: 9 tasks left</li>
<li>Claudia: 8 tasks left</li>
<li>Alexa: 5 tasks left</li>
```

> Além das contagens atualizadas, uma pessoa lendo isso pode provavelemnte dizer que nós invertermos a ordem de Alexa e Ben e inserimos Claudia entre eles. Entretanto, React é um programa de computador e não sabe da nossa intenção. Por conta disso, precisamos especificar uma propriedade *key* para cada lista de itens para diferenciar cada item dos seus irmãos. Uma opção é usar as strings `alexa`, `ben` e `claudia`. Se nós formos exibidr dados de um baco de dados, Alexa, Ben e Claudia ID's poderiam ser usados como keys:

Reviewer 2: ```suggestion
Além das contagens atualizadas, uma pessoa lendo isso pode provavelmente dizer que nós invertermos a ordem de Alexa e Ben, além de inserimos Claudia entre eles. Entretanto, React é um programa de computador e não sabe da nossa intenção. Por conta disso, precisamos especificar uma propriedade *key* para cada lista de itens para diferenciar cada item dos seus irmãos. Uma opção é usar as strings `alexa`, `ben` e `claudia`. Se nós formos exibir dados de um banco de dados, Alexa, Ben e Claudia ID's poderiam ser usados como keys:
```

```html
<li key={user.id}>{user.name}: {user.taskCount} tasks left</li>
```

> Quando uma lista é re-renderizada, React pega cada `key` de cada item da lista e procura nos items da lista anterior com as mesmas `keys`. Se a lista atual possui uma key que não existe na lista anterior, React cria um componente. Se a lista atual estiver faltando uma key que existia na lista anterior, React destroi o elemento anteiror correspondente. Se duas keys coincidirem, os componentes correspondentes são movidos. Keys dizem ao React sobre a indentidade de cada componente e o permite a manutenção entre as re-renderizações. Se a key de um componente mudar, o componente vai ser destruído e recriado com um novo estado.

Reviewer 2: ```suggestion
Quando uma lista é re-renderizada, React pega cada `key` de cada item da lista e procura nos itens da lista anterior com as mesmas `keys`. Se a lista atual possui uma key que não existe na lista anterior, React cria um componente. Se a lista atual estiver faltando uma key que existia na lista anterior, React destrói o elemento anterior correspondente. Se duas keys coincidirem, os componentes correspondentes são movidos. Keys dizem ao React sobre a identidade de cada componente e o permite a manutenção entre as re-renderizações. Se a key de um componente mudar, o componente vai ser destruído e recriado com um novo estado.
```

`key` é uma properidade especial e reservada em React (junto com `ref`, uma funcionalidade mais avançada). Quando um elemento é criado, React extrai a propriedade `key` e a guarda diretamente no elemento retornado. Mesmo que `key` pareça pertencer a `props`, ele não pode ser referenciado usando `this.props.key`. React automaticamente usa `key` para decidir qual componente deverá ser atualizado. Um componente não pode checar a sua própria `key`.

**É fortemente recomendado que você defina keys corretamente sempre que estiver construindo listas dinamicas.** Se você não tiver uma key adequada, talvez considere reestruturar seus dados.

> Se não forem especificadas keys, React vai mostrar um alerta e usar o índice do array como valor padrão para elas. Usar o índice do array como key é problematico quando nós tentamos reordenar os itens na lista ou inserir/remover novos items. Passar explicitamente `key={i}` silencia os alertas mas os mesmos problemas mencionados anteriormente ocorrerão. Não recomendamos na maioria dos casos.

Reviewer 2: ```suggestion
Se nenhuma key for especificada, o React vai mostrar um alerta e usar o índice do array como valor padrão para elas. Usar o índice do array como key é problemático quando nós tentamos reordenar os itens na lista ou inserir/remover novos itens. Passar explicitamente `key={i}` silencia os alertas mas os mesmos problemas mencionados anteriormente ocorrerão. Não recomendamos na maioria dos casos.
```

Keys não precisam ser únicas no escopo global; elas precisam ser únicas entre os componentes e seus irmãos.

### Implementando Viagem no Tempo {#implementing-time-travel}

No histórico do jogo da velha, cada jogada anterior possui um ID único associado a ele: um número sequencia da jogada. As jogadas nunca são reordenadas, deletadas ou inseridas no meio, então é seguro usar índice da jogada como key:

> No método render do componente Game, nós podemos adicionar a key como `<li key={move}>` e o alerta do React sobre keys vai disaperecer:

Reviewer 2: ```suggestion
No método render do componente Game, nós podemos adicionar a key como `<li key={move}>` e o alerta do React sobre keys vai desaparecer:
```

    });
```

**[Veja o código até este momento](https://codepen.io/gaearon/pen/PmmXRE?editors=0010)**

> Ao clicar em qualquer item da lista de botões um erro ocorrerá, pois, o método `jumpTo` não foi definido. Antes de nós o implementarmos, vamos adicionar `stepNumer` ao estado do componente Game para indicar qual passo estamos vendo naquele momento.

Reviewer 2: ```suggestion
Ao clicar em qualquer item da lista de botões um erro ocorrerá, pois, o método `jumpTo` não foi definido. Antes de nós o implementarmos, vamos adicionar `stepNumber` ao estado do componente Game para indicar qual passo estamos vendo naquele momento.
```

  }
```

> Em seguida, vamos definir o método `jumpTo` para atualizar o `stepNumber`. Nós também vamos definir `xIsNext` para true se o número que nós de `stepNumber` que estamos mudando é par:

Reviewer 2: ```suggestion
Em seguida, vamos definir o método `jumpTo` para atualizar o `stepNumber`. Nós também vamos definir `xIsNext` para true se o número que nós estamos mudando de `stepNumber` é par:
```

  }

  render() {
    // este método não foi alterado
  }
```

Nós vamos agora fazer algumas alterações no método `handleClick` do Game que vai ser ativado quando você clicar no quadrado.

> O state `stepNumber` que adicionamos reflete o movimento exibido ao usuário agora. Depois de fazer uma nova jogada, nós precisamos atualizar `stepNumber` adicionando `stepNumber: history.length` como parte do argumento `this.setState` . Isso garente que não vamos mostrar as mesmas jogadas depois de uma nova jogada feita.

Reviewer 2: ```suggestion
O state `stepNumber` que adicionamos reflete o movimento exibido ao usuário agora. Depois de fazer uma nova jogada, nós precisamos atualizar `stepNumber` adicionando `stepNumber: history.length` como parte do argumento `this.setState` . Isso garante que não vamos mostrar as mesmas jogadas depois de uma nova jogada feita.
```

> Nós vamos também substituir o uso de `this.state.history` por `this.state.history.slice(0, this.state.stepNumber + 1)`. Isso garente que se nós "voltarmos no tempo" e fizermos novas jogadas daquele momento, nós jogamos fora todo o histórico do "futuro" que agora, estaria incorreto.

Reviewer 2: ```suggestion
Nós vamos também substituir o uso de `this.state.history` por `this.state.history.slice(0, this.state.stepNumber + 1)`. Isso garante que se nós "voltarmos no tempo" e fizermos novas jogadas daquele momento, nós jogamos fora todo o histórico do "futuro" que agora, estaria incorreto.
```

  }
```

Finalmente, vamos mudar o `render` do componente Game de sempre renderizar a última jogada para renderizar a jogada atualmente selecionada de acordo com `stepNumber`:

```javascript{3}
  render() {
    const history = this.state.history;
    const current = history[this.state.stepNumber];
    const winner = calculateWinner(current.squares);

    // o resto não foi alterado
```

> Se clicarmos em qualquer passo o histórico do jogo, tabuleiro deverá imediatamente ser atualizado e mostrar como ele estava quando aquela jogada aconteceu.

Reviewer 2: ```suggestion
Se clicarmos em qualquer passo do histórico do jogo, o tabuleiro deverá imediatamente ser atualizado e mostrar como ele estava quando aquela jogada aconteceu.
```

**[Veja o código até este momento](https://codepen.io/gaearon/pen/gWWZgR?editors=0010)**

### Conclusão {#wrapping-up}

Parabéns! Você acaba de criar um jogo da velha que:

* Te permite jogar jogo da velha,
* Indica quando o jogador ganha o jogo,
* Guarda o histórico do progresso do jogo,
* Permite os jogadores rever o histórico do jogo e ver as versões anteriores do jogo no tabuleiro. 

Bom trabalho. Nós esperamos que você tenha compreendido como React funciona.

Confira aqui o resultado final: **[Resultado Final](https://codepen.io/gaearon/pen/gWWZgR?editors=0010)**.

Se você tiver um tempo extra ou quiser praticar suas novas habilidades em React, aqui estão algumas idéias para melhorar o jogo da velha e que estão listadas em ordem crescente de dificuldade:

1. Mostrar a localização de cada jogada em formato (col, row) na lista do histórico de jogadas.
2. Escurecer o atual item selecionado na lista de jogadas.
3. Reescrever o Board para criar os quadrados ao invés de tê-los fixos.
4. Adicionar um botão "toggle" que te permite ordenar as jogadas em ordem crescente ou decrescente.
5. Quando alguém ganhar, realçar os 3 quadrados usados na vitória.

> 6. When no one wins, display a message about the result being a draw.

Reviewer 2: ```suggestion
```

6. QUando ninguém ganhar, mostrar uma mensagem que o resultado foi empate.

> Ao longo deste tutorial, nós vimos conceitos React incluindo elementos, componentes, props e state. Para mais detalhes explicando cada um destes tópicos, consulte [the rest of the documentation](/docs/hello-world.html). Para apreneer mais sobre a definição de componente, consulte [`React.Component` API reference](/docs/react-component.html).

Reviewer 2: ```suggestion
Ao longo deste tutorial, nós vimos conceitos React incluindo elementos, componentes, props e state. Para mais detalhes explicando cada um destes tópicos, consulte [the rest of the documentation](/docs/hello-world.html). Para aprender mais sobre a definição de componente, consulte [`React.Component` API reference](/docs/react-component.html).
```
